mail_tracking/models/mail_tracking_email.py

```diff
@@ -0,0 +1,322 @@
+            ]).email_score()
+        return 0.
+
+    @api.model
+    def _email_score_weights(self):
+        """Default email score weights. Ready to be inherited"""
+        return {
+            'error': -50.0,
+            'rejected': -25.0,
+            'spam': -25.0,
+            'bounced': -25.0,
+            'soft-bounced': -10.0,
+            'unsub': -10.0,
+            'delivered': 1.0,
+            'opened': 5.0,
+        }
+
+    def email_score(self):
+        """Default email score algorimth. Ready to be inherited
+
+        Must return a value beetwen 0.0 and 100.0
+        - Bad reputation: Value between 0 and 50.0
+        - Unknown reputation: Value 50.0
+        - Good reputation: Value between 50.0 and 100.0
+        """
+        weights = self._email_score_weights()
+        score = 50.0
+        for tracking in self:
+            score += weights.get(tracking.state, 0.0)
+        if score > 100.0:
+            score = 100.0
+        elif score < 0.0:
+            score = 0.0
+        return score
+
+    @api.depends('recipient')
+    def _compute_recipient_address(self):
+        for email in self:
+            if email.recipient:
+                matches = re.search(r'<(.*@.*)>', email.recipient)
+                if matches:
+                    email.recipient_address = matches.group(1).lower()
+                else:
+                    email.recipient_address = email.recipient.lower()
+            else:
+                email.recipient_address = False
+
+    @api.depends('name', 'recipient')
+    def _compute_tracking_display_name(self):
+        for email in self:
+            parts = [email.name or '']
+            if email.recipient:
+                parts.append(email.recipient)
+            email.display_name = ' - '.join(parts)
+
+    @api.depends('time')
+    def _compute_date(self):
+        for email in self:
+            email.date = fields.Date.to_string(
+                fields.Date.from_string(email.time))
+
+    def _get_mail_tracking_img(self):
+        m_config = self.env['ir.config_parameter']
+        base_url = (m_config.get_param('mail_tracking.base.url') or
+                    m_config.get_param('web.base.url'))
+        path_url = (
+            'mail/tracking/open/%(db)s/%(tracking_email_id)s/blank.gif' % {
+                'db': self.env.cr.dbname,
+                'tracking_email_id': self.id,
+            })
+        track_url = urllib.parse.urljoin(base_url, path_url)
+        return (
+            '<img src="%(url)s" alt="" '
+            'data-odoo-tracking-email="%(tracking_email_id)s"/>' % {
+                'url': track_url,
+                'tracking_email_id': self.id,
+            })
+
+    @api.multi
+    def _partners_email_bounced_set(self, reason, event=None):
+        recipients = []
+        if event and event.recipient_address:
+            recipients.append(event.recipient_address)
+        else:
+            recipients = list(
```

Comment:
`list()` needed for? `mapped` already returns an iterable...

Reply:
This oddities tend to come from `2to3`. I don't know exactly attending to wich rule. Fixed anyway.